pkg/server/api/middleware/authorization.go

```diff
@@ -0,0 +1,62 @@
+package middleware
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/server/api/rpccontext"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Authorizer interface {
+	// Name returns the name the authorizer. The value may be included in logs
+	// and messages returned to callers on authorization failure.
+	Name() string
+
+	// AuthorizeCaller is called by the authorization middleware to determine
+	// if a caller is authorized. The caller is retrievable on the passed in
+	// context. On success, the method returns the (potentially embellished)
+	// context passed into the function. On failure, the method returns an
+	// error and the returned context is ignored.
+	AuthorizeCaller(ctx context.Context) (context.Context, error)
+}
+
+func WithAuthorization(authorizers map[string]Authorizer) Middleware {
+	return &authorizationMiddleware{
+		authorizers: authorizers,
+	}
+}
+
+type authorizationMiddleware struct {
+	authorizers map[string]Authorizer
+}
+
+func (m *authorizationMiddleware) Preprocess(ctx context.Context, methodName string) (context.Context, error) {
+	ctx, err := callerContextFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make(logrus.Fields)
+	if !rpccontext.CallerIsLocal(ctx) {
+		fields["caller-addr"] = rpccontext.CallerAddr(ctx).String()
+	}
+	if id, ok := rpccontext.CallerID(ctx); ok {
+		fields["caller-id"] = id.String()
+	}
+	if len(fields) > 0 {
+		ctx = rpccontext.WithLogger(ctx, rpccontext.Logger(ctx).WithFields(fields))
+	}
+
+	authorizer, ok := m.authorizers[methodName]
+	if !ok {
+		rpccontext.Logger(ctx).Error("Authorization misconfigured; this is a bug")
+		return nil, status.Errorf(codes.Internal, "authorization misconfigured")
```

Comment:
Maybe `codes.PermissionDenied` is a better code to use here? We may include the method name in the error?

Reply:
My intention here was to only return PermissionDenied if we actually determined that the caller did not have permission. In this case, this means that we (SPIRE devs) messed up and didn't set up authorization for a method. Clients may choose to do some action based on the status code and I didn't want to conflate the two. `Internal` is an indication that something is wrong outside of the callers control that they can't rectify on their own.

I didn't add the method name to the returned error because the caller presumably knows the method name they are calling? I'd be ok adding it though if you think it will help debuggability.

The logger will already have fields added with the method name.

Reply:
I think that makes sense.
Regarding including the method name, I think that I will not hurt and can help in the case of debugging this error.